### PR TITLE
Some adaptations to the dynamic data pipeline

### PR DESCRIPTION
**Pull request by francoishernandez**
1. Simplify the original "shard" logic, which is now only a "parallel corpus" logic

### CODE REVIEW

**Comment by francoishernandez** [comment]
Previous commit 8f52453 does two things:
2. Move `transform.apply` up in the codepath (in `DatasetAdapter`) --> this on its own reduces the bottleneck (not sure why)
3. Spawn `n_gpu` producer processes, which will iterate over corpus files in a strided way. This adds a small overhead in some cases, where multiple threads are not necessary, but drastically improves performance for cases where high data flow is required (e.g. high `accum_count`).

Note: the producer processes are spawned in a second loop (after the training processes) in `onmt/dynamic/train.py`. Doing everything in the same loop does not work for an obscure reason.

**Comment by francoishernandez** [comment]
Previous commit cafb9d5 removes the use of cycle, which caused a memory leak. Basically, a copy of each example was kept.
(https://stackoverflow.com/questions/16638639/why-does-pythons-itertools-cycle-need-to-create-a-copy-of-the-iterable)

**Comment by francoishernandez** [comment]
## Topic: where do we stand on reproducibility / random seed?

It seems this might be an issue here: producer processes fill the various queues in parallel --> the queues may not be filled always in the same order.

Idea: build one queue per producer per gpu (each GPU has n_producer queues).
Update: this seems to do the trick! See next commit (7f026ca).

**Comment by francoishernandez** [comment]
Note: we should add a way to keep track of where we left in the data (similar to https://github.com/OpenNMT/OpenNMT-py/pull/1826).

**Comment by francoishernandez** [comment]
Note: where do we stand on distributed (multi node) training?

**Comment by francoishernandez** [comment]
Note: looks like the execution does not end properly when `-train_steps` is reached.
Probably something to do with the added processes. Further investigation needed.
Update: solved in next commit (6a4067e).

**Comment by francoishernandez** [comment]
Note: what about the already existing noising mechanism? https://github.com/OpenNMT/OpenNMT-py/pull/1732
Some of those noise functions use specific fields that are pre-computed during preprocess, so it's not compatible as is. We should probably look at reimplemeting this in the `Transform` framework.

File: onmt/dynamic/train.py [hunk 1]
```diff
@@ -29,34 +29,43 @@ def train(opt):
...
+        producers = []
+        # This does not work if we merge with the first loop, not sure why
+        for device_id in range(nb_gpu):
+            train_iter = get_train_iter(
+                opt, dynamic=True, stride=nb_gpu, offset=device_id)
+            producer = mp.Process(target=batch_producer,
+                                  args=(train_iter, queues, semaphore, opt, device_id,),
+                                  daemon=True)
+            producers.append(producer)
+            producers[device_id].start()
```
Comment (Zenglinxiao) [inline onmt/dynamic/train.py]: Why create `nb_gpu` queues for each consumer rather than one queue for each consumer/producer?

Comment (francoishernandez) [reply onmt/dynamic/train.py]: Done in 8232b68

Comment (Zenglinxiao) [reply onmt/dynamic/train.py]: Maybe also revert the change for semaphore?

**Review by Zenglinxiao — APPROVED** [review]
Looks good. Can we merge this?

**Comment by francoishernandez** [comment]
Yes, I'll report the few remaining notes in the main PR #2 .

**Pull request merged**